File: packages/sdk-node/src/transport.ts
import { join } from "node:path";
import { appendFile, readFile, writeFile, mkdir, stat } from "node:fs/promises";
import type { ChoboConfig } from "./config.js";
import type { ChoboEvent } from "./event.js";

export interface TransportStats {
  enqueued: number;
  sent: number;
  spilled: number;
  dropped: number;
  postFailures: number;
}

/**
 * Never-blocking delivery: in-memory buffer -> interval/size-triggered flush -> batch POST.
 * Disk spill + backoff are added in the resilience task; bounded shutdown in the shutdown task.
 */
export class Transport {
  private buffer: ChoboEvent[] = [];
  private timer: NodeJS.Timeout | null = null;
  private flushing = false;
  protected closed = false;
  private readonly fetchImpl: typeof fetch;
  protected readonly spoolPath: string;
  readonly stats: TransportStats = { enqueued: 0, sent: 0, spilled: 0, dropped: 0, postFailures: 0 };
  /** Serializes all spool-file operations (the Node analogue of a lock). */
  private spoolChain: Promise<unknown> = Promise.resolve();

  /** Best-effort flush on natural process drain (safe: does not hijack signals or prevent exit). */
  private readonly onBeforeExit = (): void => {
    void this.flush();
  };

  constructor(protected readonly cfg: ChoboConfig) {
    this.fetchImpl = cfg.fetchImpl ?? fetch;
    this.spoolPath = join(cfg.spoolDir, `events-${process.pid}.jsonl`);
    this.timer = setInterval(() => { void this.flush(); }, cfg.flushIntervalMs);
    this.timer.unref();
    process.on("beforeExit", this.onBeforeExit);
  }

  /** Producer side — must be instant (no network/disk on this path). */
  enqueue(event: ChoboEvent): void {
    if (this.closed) return;
    this.stats.enqueued++;
    this.buffer.push(event);
    if (this.buffer.length > this.cfg.bufferMax) {
      const overflow = this.buffer.splice(0, this.buffer.length - this.cfg.bufferMax);
      void this.spill(overflow); // spill oldest to disk (async, non-blocking)
    }
    if (this.buffer.length >= this.cfg.flushAt) {
      void this.flush();
    }
  }

  async flush(): Promise<void> {
    if (this.flushing) return;
    this.flushing = true;
    try {
      while (this.buffer.length > 0) {
        const batch = this.buffer.splice(0, this.cfg.batchMax);
        const ok = await this.post(batch);
        if (!ok) {
          this.stats.postFailures++;
          await this.spill(batch); // failed in-memory batch -> disk, retry next pass
          break;
        }
      }
      await this.drainSpool();
    } finally {
      this.flushing = false;
    }
  }

  protected async post(events: ChoboEvent[]): Promise<boolean> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.cfg.timeoutMs);
    try {
      const res = await this.fetchImpl(this.cfg.ingestUrl, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          ...(this.cfg.ingestSecret ? { "x-chobo-secret": this.cfg.ingestSecret } : {}),
        },
        body: JSON.stringify({ events }),
        signal: controller.signal,
      });
      if (res.ok) {
        this.stats.sent += events.length;
        return true;
      }
      return false;
    } catch {
      return false;
    } finally {
      clearTimeout(timer);
    }
  }

  private withSpoolLock<T>(fn: () => Promise<T>): Promise<T> {
    const result = this.spoolChain.then(fn, fn);
    this.spoolChain = result.then(() => undefined, () => undefined);
    return result;
  }

  private async spill(events: ChoboEvent[]): Promise<void> {
    try {
      await this.withSpoolLock(async () => {
        await mkdir(this.cfg.spoolDir, { recursive: true });
        const lines = events.map((e) => JSON.stringify(e)).join("\n") + "\n";
        await appendFile(this.spoolPath, lines, "utf8");
        this.stats.spilled += events.length;
        await this.enforceSpoolCap();
      });
    } catch {
      this.stats.dropped += events.length; // disk problem — counted, never silent
    }
  }

  private async enforceSpoolCap(): Promise<void> {
    // caller holds the spool lock
    let size: number;
    try { size = (await stat(this.spoolPath)).size; } catch { return; }
    if (size <= this.cfg.maxSpoolBytes) return;
    const content = await readFile(this.spoolPath, "utf8");
    const lines = content.split("\n").filter((l) => l.length > 0);
    let bytes = Buffer.byteLength(lines.join("\n") + "\n", "utf8");
    while (lines.length && bytes > this.cfg.maxSpoolBytes) {
      lines.shift(); // drop oldest
      this.stats.dropped++;
      bytes = Buffer.byteLength(lines.length ? lines.join("\n") + "\n" : "", "utf8");
    }
    await writeFile(this.spoolPath, lines.length ? lines.join("\n") + "\n" : "", "utf8");
  }

  protected async spoolNonEmpty(): Promise<boolean> {
    try { return (await stat(this.spoolPath)).size > 0; } catch { return false; }
  }

  private async drainSpool(): Promise<void> {
    // Wrapped so a spool fs error (file removed mid-drain, permissions, etc.) can never escape as
    // an unhandled rejection from the background flusher (which could crash the host process).
    // The next flush cycle retries from whatever remains on disk. Mirrors spill()'s defensive catch.
    try {
      let lines: string[] = [];
      await this.withSpoolLock(async () => {
        let size = 0;
        try { size = (await stat(this.spoolPath)).size; } catch { return; }
        if (size === 0) return;
        const content = await readFile(this.spoolPath, "utf8");
        lines = content.split("\n").filter((l) => l.length > 0);
        await writeFile(this.spoolPath, "", "utf8"); // consume now
      });
      if (lines.length === 0) return;
      let i = 0;
      while (i < lines.length) {
        const chunk = lines.slice(i, i + this.cfg.batchMax);
        let events: ChoboEvent[];
        try { events = chunk.map((l) => JSON.parse(l) as ChoboEvent); }
        catch { i += chunk.length; continue; } // skip corrupt lines
        const ok = await this.post(events);
        if (ok) i += chunk.length;
        else break; // backoff: stop, retry next pass
      }
      const leftover = lines.slice(i);
      if (leftover.length) {
        await this.withSpoolLock(async () => {
          await appendFile(this.spoolPath, leftover.join("\n") + "\n", "utf8");
        });
      }
    } catch {
      this.stats.postFailures++; // spool fs error; swallow + count, retry next cycle
    }
  }

  async shutdown(): Promise<void> {
    this.closed = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    process.removeListener("beforeExit", this.onBeforeExit);
    // Best-effort bounded drain: stop when fully drained, or when a pass makes no progress
    // (CRM down) — in which case events remain persisted on disk, not lost.
    for (let attempt = 0; attempt < 5; attempt++) {
      const before = this.stats.sent;
      await this.flush();
      const drained = this.buffer.length === 0 && !(await this.spoolNonEmpty());
      if (drained) return;
      if (this.stats.sent === before) break; // no progress -> stop (events are on disk)
    }
  }
}

File: packages/sdk-python/src/chobo/transport.py
"""Never-blocking delivery: bounded queue -> background flusher thread -> batch POST.

This task implements in-memory delivery + flush/shutdown. Disk spill + backoff are added in
the resilience task. The flusher is a daemon thread using only stdlib urllib, so it is agnostic
to the host's event loop (works in sync and asyncio hosts, across uvicorn workers).
"""
import json
import os
import queue
import threading
import urllib.error
import urllib.request


class Transport:
    def __init__(self, ingest_url, queue_maxsize, batch_max, flush_interval,
                 spool_dir, max_spool_bytes, timeout=5.0, ingest_secret=None):
        self.ingest_url = ingest_url
        self.ingest_secret = ingest_secret
        self.batch_max = batch_max
        self.flush_interval = flush_interval
        self.spool_dir = spool_dir
        self.max_spool_bytes = max_spool_bytes
        self.timeout = timeout
        self._q = queue.Queue(maxsize=queue_maxsize)
        self._stop = threading.Event()
        self._flush_now = threading.Event()
        self._idle = threading.Event()
        self._idle.set()
        self._spool_lock = threading.Lock()
        self.stats = {"enqueued": 0, "sent": 0, "spilled": 0, "dropped": 0, "post_failures": 0}
        os.makedirs(spool_dir, exist_ok=True)
        self._spool_path = os.path.join(spool_dir, f"events-{os.getpid()}.jsonl")
        self._thread = threading.Thread(target=self._run, name="chobo-flusher", daemon=True)
        self._thread.start()

    # ---- producer side (called on the business thread; must be instant) ----
    def enqueue(self, event):
        self.stats["enqueued"] += 1
        self._idle.clear()
        try:
            self._q.put_nowait(event)
        except queue.Full:
            self._spill([event])

    # ---- consumer side (flusher thread) ----
    def _run(self):
        while not self._stop.is_set():
            fired = self._flush_now.wait(timeout=self.flush_interval)
            if fired:
                self._flush_now.clear()
            self._drain_once()
        self._drain_once()  # final drain on shutdown

    def _take_batch(self):
        batch = []
        for _ in range(self.batch_max):
            try:
                batch.append(self._q.get_nowait())
            except queue.Empty:
                break
        return batch

    def _spill(self, events):
        with self._spool_lock:
            with open(self._spool_path, "a", encoding="utf-8") as f:
                for e in events:
                    f.write(json.dumps(e, ensure_ascii=False) + "\n")
            self.stats["spilled"] += len(events)
            self._enforce_spool_cap()

    def _enforce_spool_cap(self):
        # caller holds _spool_lock
        try:
            if os.path.getsize(self._spool_path) <= self.max_spool_bytes:
                return
        except OSError:
            return
        with open(self._spool_path, "r", encoding="utf-8") as f:
            lines = f.readlines()
        size = sum(len(x.encode("utf-8")) for x in lines)
        while lines and size > self.max_spool_bytes:
            dropped = lines.pop(0)              # drop OLDEST (never silent)
            size -= len(dropped.encode("utf-8"))
            self.stats["dropped"] += 1
        with open(self._spool_path, "w", encoding="utf-8") as f:
            f.writelines(lines)

    def _spool_nonempty(self):
        try:
            return os.path.getsize(self._spool_path) > 0
        except OSError:
            return False

    def _drain_spool(self):
        # Consume the spool atomically: read all lines AND truncate under a single lock hold,
        # so a concurrent _spill() during the (unlocked) POST window cannot be erased by a later
        # truncate. Unsent leftover is re-appended afterwards (append-only, lock-guarded).
        with self._spool_lock:
            if not self._spool_nonempty():
                return
            with open(self._spool_path, "r", encoding="utf-8") as f:
                lines = [ln for ln in f.read().splitlines() if ln.strip()]
            open(self._spool_path, "w", encoding="utf-8").close()  # consume: truncate now
        i = 0
        while i < len(lines):
            chunk = lines[i:i + self.batch_max]
            try:
                events = [json.loads(x) for x in chunk]
            except json.JSONDecodeError:
                i += len(chunk)
                continue
            if self._post(events):
                i += len(chunk)
            else:
                break
        leftover = lines[i:]
        if leftover:
            with self._spool_lock:
                with open(self._spool_path, "a", encoding="utf-8") as f:
                    for ln in leftover:
                        f.write(ln + "\n")

    def _drain_once(self):
        self._idle.clear()
        while True:
            batch = self._take_batch()
            if not batch:
                break
            if not self._post(batch):
                self.stats["post_failures"] += 1
                self._spill(batch)     # failed in-memory batch -> disk, retry next cycle
                break
        self._drain_spool()
        if self._q.empty() and not self._spool_nonempty():
            self._idle.set()

    def _post(self, events):
        body = json.dumps({"events": events}, ensure_ascii=False).encode("utf-8")
        headers = {"Content-Type": "application/json"}
        if self.ingest_secret:
            headers["x-chobo-secret"] = self.ingest_secret
        req = urllib.request.Request(self.ingest_url, data=body, headers=headers, method="POST")
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                if 200 <= resp.status < 300:
                    self.stats["sent"] += len(events)
                    return True
                return False
        except (urllib.error.URLError, OSError):
            return False

    # ---- lifecycle ----
    def flush(self, timeout=5.0):
        self._idle.clear()
        self._flush_now.set()
        return self._idle.wait(timeout=timeout)

    def shutdown(self, timeout=5.0):
        self.flush(timeout=timeout)
        self._stop.set()
        self._flush_now.set()
        self._thread.join(timeout=timeout)
